kernel/src/protocols/vtpm.rs

```diff
@@ -0,0 +1,220 @@
+/// vTPM platform commands (SVSM spec, section 8.1 - SVSM_VTPM_QUERY)
+///
+/// The platform commmand values follow the values used by the
+/// Official TPM 2.0 Reference Implementation by Microsoft.
+///
+/// `ms-tpm-20-ref/TPMCmd/Simulator/include/TpmTcpProtocol.h`
+#[repr(u32)]
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum TpmPlatformCommand {
+    SendCommand = 8,
+}
+
+impl TryFrom<u32> for TpmPlatformCommand {
+    type Error = SvsmReqError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        let cmd = match value {
+            x if x == TpmPlatformCommand::SendCommand as u32 => TpmPlatformCommand::SendCommand,
+            other => {
+                log::warn!("Failed to convert {} to a TPM platform command", other);
+                return Err(SvsmReqError::invalid_parameter());
+            }
+        };
+
+        Ok(cmd)
+    }
+}
+
+fn vtpm_platform_commands_supported_bitmap() -> u64 {
+    let mut bitmap: u64 = 0;
+    let vtpm = vtpm_get_locked();
+
+    for cmd in vtpm.get_supported_commands() {
+        bitmap |= 1u64 << *cmd as u32;
+    }
+
+    bitmap
+}
+
+fn is_vtpm_platform_command_supported(cmd: TpmPlatformCommand) -> bool {
+    let vtpm = vtpm_get_locked();
+    vtpm.get_supported_commands().iter().any(|x| *x == cmd)
+}
+
+const SEND_COMMAND_REQ_INBUF_SIZE: usize = PAGE_SIZE - 9;
+
+// vTPM protocol services (SVSM spec, table 14)
+const SVSM_VTPM_QUERY: u32 = 0;
+const SVSM_VTPM_COMMAND: u32 = 1;
+
+/// TPM_SEND_COMMAND request structure (SVSM spec, table 16)
+#[derive(Clone, Copy, Debug)]
+#[repr(C, packed)]
+struct TpmSendCommandRequest {
+    /// MSSIM platform command ID
+    command: u32,
+    /// Locality usage for the vTPM is not defined yet (must be zero)
+    locality: u8,
+    /// Size of the input buffer
+    inbuf_size: u32,
+    /// Input buffer that contains the TPM command
+    inbuf: [u8; SEND_COMMAND_REQ_INBUF_SIZE],
+}
+
+impl TpmSendCommandRequest {
+    pub fn send(&self) -> Result<Vec<u8>, SvsmReqError> {
+        // TODO: Before implementing locality, we need to agree what it means
+        // to the platform
+        if self.locality != 0 {
+            return Err(SvsmReqError::invalid_parameter());
+        }
+
+        let mut length = self.inbuf_size as usize;
+
+        let tpm_cmd = self
+            .inbuf
+            .get(..length)
+            .ok_or_else(SvsmReqError::invalid_parameter)?;
+        let mut buffer: Vec<u8> = Vec::with_capacity(SEND_COMMAND_RESP_OUTBUF_SIZE);
+        buffer.extend_from_slice(tpm_cmd);
+
+        // The buffer slice must be large enough to hold the TPM command response
+        unsafe { buffer.set_len(buffer.capacity()) };
+
+        let vtpm = vtpm_get_locked();
+        vtpm.send_tpm_command(buffer.as_mut_slice(), &mut length, self.locality)?;
+
+        unsafe {
+            if length > buffer.capacity() {
+                return Err(SvsmReqError::invalid_request());
+            }
+            buffer.set_len(length);
+        }
```

Comment:
Again, we cannot set the length to capacity because that could include allocated but uninitialized memory (note that capacity might be greater than `SEND_COMMAND_RESP_OUTBUF_SIZE`, we are only guaranteed that's the minimum).

With my previous suggestion we filled up buffer until `SEND_COMMAND_RESP_OUTBUF_SIZE` with zeroes, so we can simply check that the new length is equal or smaller, and then truncate, avoiding unsafe completely.

```suggestion
        if length > buffer.len() {
            return Err(SvsmReqError::invalid_request());
        }
        buffer.truncate(length);
```